### PR TITLE
Latent retries

### PR DESCRIPTION
**Pull request by bergden-resonai**
What this PR doesn't do:
Test that predecessors are indeed skipped
Accumulate prints and clearing the clutter (it only gives you a list of failed targets).

### CODE REVIEW

File: yabt/cli.py [hunk 1]
```diff
@@ -81,6 +81,8 @@ def make_parser(project_config_file: str) -> configargparse.ArgumentParser:
         # PARSER.add('--config', is_config_file=True, help='Config file path')
         PARSER.add('--artifacts-metadata-file',
                    help='Output file to write artifacts metadata to')
+        PARSER.add('--continue-after-fail', default=False, action='store_true',
+                   help='If a target fails continue independant targets')
```
Comment (itamaro) [inline yabt/cli.py]:
```suggestion
                   help='If a target fails continue independent targets')
```

File: yabt/buildcontext.py [hunk 2]
```diff
@@ -453,37 +503,14 @@ def build(target: Target):
...
-                            except Exception:
-                                target.info['fail_count'] += 1
-                                if attempts <= target.info['fail_count']:
-                                    raise
-                            else:
-                                break
+                        test_start = time()
+                        self.test_target(target)
                         target.info['test_time'] = time() - test_start
                         logger.info(
```
Comment (itamaro) [inline yabt/buildcontext.py]: is it still relevant to log `fail_count` here given the changes?

File: yabt/buildcontext.py [hunk 3]
```diff
@@ -495,13 +522,23 @@ def build(target: Target):
...
-                target.fail()
-                fatal('`{}\': {}', target.name, ex)
+                target.info['fail_count'] += 1
+                default_attempts = 1
+                if 'testable' in target.tags:
+                    default_attempts = self.conf.test_attempts
+                attempts = max(target.props.attempts, default_attempts)
+                if attempts > target.info['fail_count']:
+                    target.retry()
+                    pass
```
Comment (itamaro) [inline yabt/buildcontext.py]: this `pass` doesn't do anything. what was the intention?

File: yabt/buildcontext.py [hunk 4]
```diff
@@ -253,6 +255,49 @@ def done_notifier():
...
+            """Return a callable "fail" notifier to
+               report a target as failed after all retries."""
+
+            def fail_notifier(ex):
+                """Mark target as failed, taking it and decendants
+                   out from the queue"""
+                if graph_copy.has_node(target.name):
+                    self.failed_nodes.append(target.name)
+                    affected_nodes = [target.name]
+                    # removing all predecessors recursively
```
Comment (itamaro) [inline yabt/buildcontext.py]: no need to manually implement the "recursive get affected nodes" algorithm - just use the built-in one (see https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.ancestors.html).

also see https://github.com/resonai/ybt/blob/master/yabt/graph.py#L340 (but I think you want "ancestors" not "descendants")

File: yabt/buildcontext.py [hunk 5]
```diff
@@ -253,6 +255,49 @@ def done_notifier():
...
+        def make_retry_callback(target: Target):
+            """Return a callable "retry" notifier to
+               report a target as in need of retry.
+               Currently for tests we rebuild the target
+               when it's not necessary."""
+
+            def retry_notifier():
+                """Mark target as retry, re-entering node to end of queue"""
+                if graph_copy.has_node(target.name):
+                    ready_nodes.append(target.name)
```
Comment (itamaro) [inline yabt/buildcontext.py]: also verify that it's not already in `ready_nodes`

File: yabt/buildcontext.py [hunk 6]
```diff
@@ -495,13 +522,23 @@ def build(target: Target):
                         logger.info(
                             'Target {} test cached - skipping test run',
                             target.name)
+                built_targets.add(target.name)
+                target.done()
```
[the remark this answers is not in the document]
Comment (bergden-resonai) [reply yabt/buildcontext.py]: I though "build_cached" should avoid re-building the target.
In the current design as I understood there's a strong coupling between a test and its build. "built_target" refers to the couple where "build_cache" refers just to the build.

File: yabt/buildcontext.py (hunk 2 above)
Comment (bergden-resonai) [reply yabt/buildcontext.py]: Yeah, that's the only place where we document the flakiness of a test.
I can move this print to the bottom. But this is not a failed target, it's a target the passed on its n-th trial.

File: yabt/buildcontext.py (hunk 5 above)
Comment (bergden-resonai) [reply yabt/buildcontext.py]: If it just failed it shouldn't be in the ready_nodes, why check it?
I don't see a flow where we might push something twice to the queue.

File: yabt/buildcontext.py (hunk 4 above)
Comment (bergden-resonai) [reply yabt/buildcontext.py]: This creates following error in run-time:
AttributeError: 'DiGraph' object has no attribute 'ancestors'
https://networkx.github.io/documentation/networkx-1.10/reference/classes.digraph.html

File: yabt/buildcontext.py [hunk 7]
```diff
@@ -524,4 +561,11 @@ def build_in_pool(seq):
                     self.conf.flavor, self.conf.jobs)
         # main pass: build rest of the graph
         build_in_pool(self.target_iter())
-        logger.info('Finished building target graph successfully')
+        if self.failed_nodes:
+            msg = ('Finished building target graph with fails: \n{}\n'
+                   'which caused the following to skip: \n{}'.format(
+                   self.failed_nodes, self.skipped_nodes))
+            print(msg)
```
Comment (bergden-resonai) [inline yabt/buildcontext.py]: @itamaro Can't use "fatal" here cause we're not in an exception context. Which fails only on python 3.4
WDYT?

File: yabt/buildcontext.py (hunk 4 above)
Comment (itamaro) [reply yabt/buildcontext.py]: it's not a method on the DiGraph. it's a function in the dag algorithms module - you pass the graph as an argument

File: yabt/buildcontext.py (hunk 2 above)
Comment (itamaro) [reply yabt/buildcontext.py]: your call

File: yabt/buildcontext.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (itamaro) [reply yabt/buildcontext.py]: I think you're right. this is confusing. can you test it? (not a unittest, just manually)

File: yabt/utils.py [hunk 8]
```diff
@@ -52,6 +52,16 @@ def fatal(msg, *args, **kwargs):
     sys.exit(1)
```
Comment (itamaro) [inline yabt/utils.py]: replace the last 3 lines of this function with a call to `fatal_noexc`

File: yabt/graph.py [hunk 9]
```diff
@@ -340,3 +340,8 @@ def topological_sort(graph: networkx.DiGraph):
 def get_descendants(graph: networkx.DiGraph, source):
     """Return all nodes reachable from `source` in `graph`."""
     return dag.descendants(graph, source)
+
+
+def get_ancestors(graph: networkx.DiGraph, source):
+    """Return all nodes that direct to `source` in `graph`."""
```
Comment (itamaro) [inline yabt/graph.py]: wording not clear in this docstring

File: yabt/buildcontext.py [hunk 10]
```diff
@@ -524,4 +558,10 @@ def build_in_pool(seq):
                     self.conf.flavor, self.conf.jobs)
         # main pass: build rest of the graph
         build_in_pool(self.target_iter())
-        logger.info('Finished building target graph successfully')
+        if self.failed_nodes:
+            msg = ('Finished building target graph with fails: \n{}\n'
```
Comment (itamaro) [inline yabt/buildcontext.py]: why intermediate variable? also, `fatal_noexc` can take a format string with format variables as args (like logger)

File: yabt/buildcontext.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (itamaro) [reply yabt/buildcontext.py]: cool

File: yabt/graph.py (hunk 9 above)
Comment (bergden-resonai) [reply yabt/graph.py]: def get_ancestors(graph: networkx.DiGraph, child):
    """Return all nodes that have 'child' as a descendant in 'graph'"""

?

File: yabt/utils.py (hunk 8 above)
Comment (bergden-resonai) [reply yabt/utils.py]: Done

File: yabt/buildcontext.py (hunk 10 above)
Comment (bergden-resonai) [reply yabt/buildcontext.py]: Done

**Pull request merged**